packages/@glimmer/runtime/lib/compiled/opcodes/vm.ts

```diff
@@ -34,28 +34,17 @@ APPEND_OPCODES.add(Op.Constant, (vm, { op1: other }) => {
 
 APPEND_OPCODES.add(Op.Primitive, (vm, { op1: primitive }) => {
   let stack = vm.stack;
-  let flag = primitive & 7; // 111
-  let value = primitive >> 3;
-
-  switch (flag) {
-    case PrimitiveType.NUMBER:
-      stack.push(value);
-      break;
-    case PrimitiveType.FLOAT:
-      stack.push(vm[CONSTANTS].getNumber(value));
-      break;
-    case PrimitiveType.STRING:
-      stack.push(vm[CONSTANTS].getString(value));
-      break;
-    case PrimitiveType.BOOLEAN_OR_VOID:
-      stack.pushRaw(primitive);
-      break;
-    case PrimitiveType.NEGATIVE:
-      stack.push(vm[CONSTANTS].getNumber(value));
-      break;
-    case PrimitiveType.BIG_NUM:
-      stack.push(vm[CONSTANTS].getNumber(value));
-      break;
+  if (primitive < 0) {
+    // is handle
+    let handle = ~primitive;
+    if (handle < 1073741824) {
+      stack.push(vm[CONSTANTS].getString(handle));
+    } else {
+      stack.push(vm[CONSTANTS].getNumber(handle - 1073741824));
```

Comment:
Probably should put this in enum type.

Reply:
will do, this just divides the handle space in 2.  in the constants for arg operands we have 2.  on the program side there is just one array for all non directly encodable things `writeJs`,  this is why that part of the code isn't shared.